src/main/java/com/google/devtools/build/lib/analysis/ConfiguredRuleClassProvider.java

```diff
@@ -899,7 +905,7 @@ public ConstraintSemantics getConstraintSemantics() {
   }
 
   /** Returns all skylark objects in global scope for this RuleClassProvider. */
-  public Map<String, Object> getTransitiveGlobalBindings() {
+  public Map<Identifier, Object> getTransitiveGlobalBindings() {
```

Comment:
It is the job of the static resolver (validation) pass to resolve each identifier to the logical variable to which it refers.  For a global variable, you could use a small integer index into a flat array, or the string name of the variable (just as the existing code does) since names within a module are unique. It makes no sense to use Identifiers in the representation of the environment: identifiers are not canonical.

I recommend you first read and understand how the Go implementation resolves identifiers and represents the environment, as it is the clearest expression of the necessary algorithm.